File: day8_part1_2016.java

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Solution {

    private static final int screenWidth = 50;
    private static final int screenHeight = 6;

    public static void main(String[] args) {
        boolean[][] screen = new boolean[screenHeight][screenWidth];

        try {
            File file = new File("input.txt");
            Scanner scanner = new Scanner(file);

            while (scanner.hasNextLine()) {
                processInstruction(scanner.nextLine(), screen);
            }

            System.out.println(countLitPixels(screen));

            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    private static void processInstruction(String instruction, boolean[][] screen) {
        Pattern rectPattern = Pattern.compile("rect (\\d+)x(\\d+)");
        Pattern rotateRowPattern = Pattern.compile("rotate row y=(\\d+) by (\\d+)");
        Pattern rotateColumnPattern = Pattern.compile("rotate column x=(\\d+) by (\\d+)");

        Matcher rectMatcher = rectPattern.matcher(instruction);
        Matcher rotateRowMatcher = rotateRowPattern.matcher(instruction);
        Matcher rotateColumnMatcher = rotateColumnPattern.matcher(instruction);

        if (rectMatcher.find()) {
            int a = Integer.parseInt(rectMatcher.group(1));
            int b = Integer.parseInt(rectMatcher.group(2));
            rect(screen, a, b);
        } else if (rotateRowMatcher.find()) {
            int a = Integer.parseInt(rotateRowMatcher.group(1));
            int b = Integer.parseInt(rotateRowMatcher.group(2));
            rotateRow(screen, a, b);
        } else if (rotateColumnMatcher.find()) {
            int a = Integer.parseInt(rotateColumnMatcher.group(1));
            int b = Integer.parseInt(rotateColumnMatcher.group(2));
            rotateColumn(screen, a, b);
        }
    }

    private static void rect(boolean[][] screen, int a, int b) {
        for (int y = 0; y < b; y++) {
            for (int x = 0; x < a; x++) {
                screen[y][x] = true;
            }
        }
    }

    private static void rotateRow(boolean[][] screen, int row, int shift) {
        boolean[] temp = new boolean[screenWidth];
        for (int i = 0; i < screenWidth; i++) {
            temp[(i + shift) % screenWidth] = screen[row][i];
        }
        screen[row] = temp;
    }

    private static void rotateColumn(boolean[][] screen, int col, int shift) {
        boolean[] temp = new boolean[screenHeight];
        for (int i = 0; i < screenHeight; i++) {
            temp[(i + shift) % screenHeight] = screen[i][col];
        }
        for (int i = 0; i < screenHeight; i++) {
            screen[i][col] = temp[i];
        }
    }

    private static int countLitPixels(boolean[][] screen) {
        int count = 0;
        for (boolean[] row : screen) {
            for (boolean pixel : row) {
                if (pixel) {
                    count++;
                }
            }
        }
        return count;
    }
}

File: fortran90/day8_part1_2016.f90

program screen_pixels
  implicit none
  integer, parameter :: screenWidth = 50
  integer, parameter :: screenHeight = 6
  logical :: screen(screenHeight, screenWidth)
  character(len=100) :: line
  integer :: i, j, a, b, count
  
  open(unit=10, file="input.txt", status="old")

  screen = .false.

  do
    read(10, '(A)', iostat=i) line
    if (i /= 0) exit
    call processInstruction(line, screen)
  end do

  close(10)

  count = 0
  do i = 1, screenHeight
    do j = 1, screenWidth
      if (screen(i,j)) count = count + 1
    end do
  end do

  print *, count

contains

  subroutine processInstruction(instruction, screen)
    implicit none
    character(len=*), intent(in) :: instruction
    logical, intent(inout) :: screen(screenHeight, screenWidth)
    integer :: a, b
    
    if (instruction(1:4) == "rect") then
      read(instruction(6:),'(I2,1X,A1,1X,I2)') a, line, b
      call rect(screen, a, b)
    else if (instruction(1:10) == "rotate row") then
      read(instruction(14:),'(I1,3X,A2,1X,I2)') a, line, b
      call rotateRow(screen, a, b)
    else if (instruction(1:13) == "rotate column") then
      read(instruction(17:),'(I2,3X,A2,1X,I2)') a, line, b
      call rotateColumn(screen, a, b)
    end if
  end subroutine

  subroutine rect(screen, a, b)
    implicit none
    logical, intent(inout) :: screen(screenHeight, screenWidth)
    integer, intent(in) :: a, b
    integer :: i, j
    do i = 1, b
      do j = 1, a
        screen(i,j) = .true.
      end do
    end do
  end subroutine

  subroutine rotateRow(screen, row, shift)
    implicit none
    logical, intent(inout) :: screen(screenHeight, screenWidth)
    integer, intent(in) :: row, shift
    logical :: temp(screenWidth)
    integer :: i
    temp = screen(row,:)
    do i = 1, screenWidth
      screen(row, mod(i+shift-1, screenWidth)+1) = temp(i)
    end do
  end subroutine

  subroutine rotateColumn(screen, col, shift)
    implicit none
    logical, intent(inout) :: screen(screenHeight, screenWidth)
    integer, intent(in) :: col, shift
    logical :: temp(screenHeight)
    integer :: i
    temp = screen(:,col)
    do i = 1, screenHeight
      screen(mod(i+shift-1, screenHeight)+1, col) = temp(i)
    end do
  end subroutine

end program screen_pixels
